torch/_dynamo/output_graph.py

```diff
@@ -2229,6 +2199,296 @@ def maybe_lift_tracked_freevar_to_input(self, arg):
+                    )
+                    assert isinstance(s, torch.SymInt)  # to make linter happy
+                    tracer.bound_symbols[s.node.expr] = lazy_proxy
+        elif t.layout is torch.sparse_coo:
+            self._lazy_bind_unbacked_symbols_in_tensors(tracer, t._indices(), t_proxy)
+            self._lazy_bind_unbacked_symbols_in_tensors(tracer, t._values(), t_proxy)
+        elif t.layout in {torch.sparse_csr, torch.sparse_bsr}:
+            self._lazy_bind_unbacked_symbols_in_tensors(
+                tracer, t.crow_indices(), t_proxy
+            )
+            self._lazy_bind_unbacked_symbols_in_tensors(
+                tracer, t.col_indices(), t_proxy
+            )
+        elif t.layout in {torch.sparse_csc, torch.sparse_bsc}:
+            self._lazy_bind_unbacked_symbols_in_tensors(
+                tracer, t.ccol_indices(), t_proxy
+            )
+            self._lazy_bind_unbacked_symbols_in_tensors(
+                tracer, t.row_indices(), t_proxy
+            )
+        if is_traceable_wrapper_subclass(t):
+            attrs, ctx = t.__tensor_flatten__()
+            for attr in attrs:
+                inner_t = getattr(t, attr)
+                self._lazy_bind_unbacked_symbols_in_tensors(
+                    tracer, inner_t, getattr(t_proxy, attr)
+                )
+
+    # NOTE: [Auto lift basic free symbols when create_graph_input]
+    # Whenever we call create_graph_input, we try to also lift the basic symbols in example values
+    # as graph input.
+    # This applies to both top-level graph and subgraphs in higher order ops.
+    # This has several cases:
+    #  1. When create_graph_input for a tensor that has symbolic shapes,
+    #     we look for basic symbols in its size and stride, we check if the symbol is bound
+    #     in current graph (i.e. bound_symbols), it it's not bound, we'll create a placeholder
+    #     for it then recursively check its parent, creates ph if not bound.
+    #     Every tracer maintains a mapping (i.e. lifted_freevars)
+    #     that maps from parent proxy to proxy in current tracer for the symbol.
+    #  2. When create_graph_input for a tensor with unbacked symbolic shapes,
+    #     Backed symbols all come from inputs's symbolic shape. But unbacked symbols
+    #     can be created while tracing. So we use lazy_bind_unbacked_symbols will intercept
+    #     at wrap_fx_proxy, and try to bind the unbacked symbols immediately after they're
+    #     created.
+    #  3. subgraph will also lifted basic symbols in compound exprs.
+    #     For example, if an input to subgraph takes size [s1+s2//8], we'll look for the
+    #     the free symbols in the sizes and lift as inputs similar to 1 in _lift_symbols_in_symint)
+    # Do not call this function direclty. This is used in create_graph_input for
+    # automatically lifting the free basic symbols in example_value.
+    def _lift_basic_symbols(self, example_value, src: Optional[Source]):
```

Comment:
why is this a helper function? It gets called once. might be simpler if we just inline this

Reply:
Yeah, I was thinking about torchbind object might also want to bind symbols in it. But i'm ok to move it out